src/data/mv_polynomial/equiv.lean

```diff
@@ -104,37 +104,74 @@ def punit_ring_equiv : mv_polynomial punit R ≃+* polynomial R :=
   map_mul'  := λ _ _, eval₂_mul _ _,
   map_add'  := λ _ _, eval₂_add _ _ }
 
-/-- The ring isomorphism between multivariable polynomials induced by an equivalence
-of the variables.  -/
+/-- If  `e : A ≃+* B` is a rings isomorphism and `e_var : S₁ ≃ S₂` is an isomorphism of types,
+the induced isomorphism `mv_polynomial S₁ A ≃+* mv_polynomial S₂ B`. -/
 @[simps]
-def ring_equiv_of_equiv (e : S₁ ≃ S₂) : mv_polynomial S₁ R ≃+* mv_polynomial S₂ R :=
-{ to_fun    := rename e,
-  inv_fun   := rename e.symm,
-  left_inv  := λ p, by simp only [rename_rename, (∘), e.symm_apply_apply]; exact rename_id p,
-  right_inv := λ p, by simp only [rename_rename, (∘), e.apply_symm_apply]; exact rename_id p,
-  map_mul'  := (rename e).map_mul,
-  map_add'  := (rename e).map_add }
+def ring_equiv_congr {A : Type u} {B : Type x} [comm_semiring A] [comm_semiring B] (e : A ≃+* B)
+  (e_var : S₁ ≃ S₂) : mv_polynomial S₁ A ≃+* mv_polynomial S₂ B :=
+{ to_fun    := rename e_var ∘ map (e : A →+* B),
+  inv_fun   := map (e.symm : B →+* A) ∘ rename e_var.symm,
```

Comment:
Given the definition is just a composition, why is this an improvement? It makes the proofs longer, and the expressions more complex.

Reply:
That is to say, why is `p.rename_and_map e_var e` a better API than `(p.rename e_var).map e`?

Reply:
Those are still available, right? This is just to get a general equiv, which can be specialized later.

Reply:
Yes they are still available, they are called `left` and `right`.

Reply:
I guess this feels to me like defining:
```
structure nat_wrap := (val : ℕ)

namespace nat_wrap

instance : has_one nat_wrap := ⟨⟨1⟩⟩
instance : has_zero nat_wrap := ⟨⟨0⟩⟩

def mul_add : ∀ (x y z : nat_wrap), nat_wrap
| ⟨a⟩ ⟨b⟩ ⟨c⟩ := ⟨a * b + c⟩

instance : has_add nat_wrap := ⟨⟨λ a b, mul_add 1 a b⟩⟩
instance : has_mul nat_wrap := ⟨⟨λ a b, mul_add a b 0⟩⟩

end nat_wrap
```

I mean, yes, it works - but it seems weird to me to define a complex thing in terms of a composition of simple things, and then break it back into pieces, rather than define the simple things separately.

Having said that, I think the `comap` stuff makes it awkward to do things the way I was proposing, as we don't currently have `alg_equiv.comap`.

Reply:
No, I mean the alg_equiv ` { to_fun := e.to_alg_hom.comap S, inv_fun := e.symm.to_alg_hom.comap S, ...}`

Reply:
Specifically, I mean the definition I add in #6548